### PR TITLE
Make compression use a single job rather than a job-per-ht

### PR DESCRIPTION
**Pull request by cevian**
This change effects only TimescaleDB 2.0 for now. Instead of
having one compression job per hypertabale, we switch to a model
of one compression job that acts on all hypertables.
This has 2 main advantages:
1) It scales better with number of hypertables
2) Allows users to run job through user defined actions inside the DB
   OR use CRON, allowing for more flexibility.

### CODE REVIEW

**Review by Harkishen-Singh — APPROVED** [review]
AFAIK, LGTM.

**Review by antekresic — APPROVED** [review]
:+1:

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 1]
```diff
@@ -1837,13 +1889,15 @@ CALL execute_everywhere($ee$
...
+            IF NOT transactional THEN
+              COMMIT;
+            END IF;
         END LOOP;
     END;
     $$ LANGUAGE PLPGSQL;
 $ee$);
 
-CREATE OR REPLACE PROCEDURE SCHEMA_CATALOG.decompress_chunks_after(metric_table NAME, min_time TIMESTAMPTZ)
+CREATE OR REPLACE PROCEDURE SCHEMA_CATALOG.decompress_chunks_after(metric_table NAME, min_time TIMESTAMPTZ, transactional BOOLEAN = false)
```
Comment (JLockerman) [inline pkg/pgmodel/migrations/sql/idempotent/base.sql:1900]: do we ever call this with `transactional => true`?

File: pkg/pgmodel/migrations/sql/preinstall/005-install_uda.sql [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+--add 2 jobs executing every 30 min by default for timescaledb 2.0
+DO $$
+BEGIN
+    IF SCHEMA_CATALOG.get_timescale_major_version() >= 2 THEN
+       PERFORM add_job('SCHEMA_CATALOG.execute_maintenance_job', '30 min');
+       PERFORM add_job('SCHEMA_CATALOG.execute_maintenance_job', '30 min');
+    END IF;
+END
+$$;
```
Comment (JLockerman) [inline pkg/pgmodel/migrations/sql/preinstall/005-install_uda.sql:25]: needs a trailing `\n`

Comment (JLockerman) [inline pkg/pgmodel/migrations/sql/preinstall/005-install_uda.sql:25]: should we add a check that there's at least 1 job to the `idempotent` scripts? I'm concercned about how we'll handle the eventual upgrades from Timescale 1 to 2

File: pkg/pgmodel/migrations/sql/versions/dev/0.1.4-dev/1-change_compression_job.sql [hunk 3]
```diff
@@ -0,0 +1,6 @@
+ALTER TABLE SCHEMA_CATALOG.metric
+    ADD COLUMN delay_compression_until TIMESTAMPTZ DEFAULT NULL;
+
+    DROP PROCEDURE IF EXISTS SCHEMA_CATALOG.decompress_chunks_after(NAME, TIMESTAMPTZ);
+    DROP PROCEDURE IF EXISTS SCHEMA_CATALOG.do_decompress_chunks_after(NAME, TIMESTAMPTZ);
+    DROP PROCEDURE IF EXISTS SCHEMA_CATALOG.compression_job(INT, JSONB);
```
Comment (JLockerman) [inline pkg/pgmodel/migrations/sql/versions/dev/0.1.4-dev/1-change_compression_job.sql:6]: trailing `\n`

**Review by JLockerman — APPROVED** [review]
It looks reasonable, though it could use some comments explaining the "why"s for some of the things, in particular the locking

File: pkg/pgmodel/migrations/sql/idempotent/base.sql (hunk 1 above)
Comment (cevian) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1900]: Yes, we call it when turning compression off

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 4]
```diff
@@ -1837,13 +1889,15 @@ CALL execute_everywhere($ee$
                 PERFORM decompress_chunk(format('%I.%I', chunk_row.schema_name, chunk_row.table_name)::regclass);
             END IF;
 
-            COMMIT;
+            IF NOT transactional THEN
```
[the remark this answers is not in the document]
Comment (cevian) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1892]: function param

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 5]
```diff
@@ -1798,7 +1845,8 @@ LANGUAGE PLPGSQL;
 CALL execute_everywhere($ee$
     --Decompression should take place in a procedure because we don't want locks held across
     --decompress_chunk calls since that function takes some heavier locks at the end.
-    CREATE OR REPLACE PROCEDURE SCHEMA_CATALOG.do_decompress_chunks_after(metric_table NAME, min_time TIMESTAMPTZ)
+    --Thus, transactional parameter should usually be false
+    CREATE OR REPLACE PROCEDURE SCHEMA_CATALOG.do_decompress_chunks_after(metric_table NAME, min_time TIMESTAMPTZ, transactional BOOLEAN = false)
```
[the remark this answers is not in the document]
Comment (cevian) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1849]: not when turning compression off

File: pkg/pgmodel/migrations/sql/idempotent/base.sql [hunk 6]
```diff
@@ -1811,7 +1859,11 @@ CALL execute_everywhere($ee$
 
         SELECT d.* INTO STRICT dimension_row FROM _timescaledb_catalog.dimension d WHERE hypertable_id = hypertable_row.id ORDER BY id LIMIT 1;
 
-        SELECT _timescaledb_internal.time_to_internal(min_time) INTO STRICT min_time_internal;
+        IF min_time = timestamptz '-Infinity' THEN
+            min_time_internal := -9223372036854775808;
+        ELSE
+           SELECT _timescaledb_internal.time_to_internal(min_time) INTO STRICT min_time_internal;
+        END IF;
```
[the remark this answers is not in the document]
Comment (cevian) [reply pkg/pgmodel/migrations/sql/idempotent/base.sql:1866]: when turning compression off

File: pkg/pgmodel/migrations/sql/preinstall/005-install_uda.sql (hunk 2 above)
Comment (cevian) [reply pkg/pgmodel/migrations/sql/preinstall/005-install_uda.sql:25]: I dunno what to do about that. Having 0 jobs is totally reasonable if using cron. We'll need some manual intervention.

**Pull request merged**